### PR TITLE
Avoid erts_cmp jump in atom and int comparisons

### PR DESCRIPTION
**Pull request by josevalim**
Given the two possible definitions for the same
function:

    check(X) when X >= 0, X <= 20 -> true.

and:

    check(0) -> true;
    check(1) -> true;
    ...
    check(20) -> true.

@nox has originally noticed that perfoming lt and ge
guard tests (the first) were drastically slower than
matching on values (the second).

Further investigation revealed that most of the cost
was in jumping to the erts_cmp function. This patch
brings the operations already inlined in erts_cmp
into the emulator, removing the jump cost.

Notice floats were not inlined, as they are not
constants and handled by erts_cmp_compound anyway.

This patch also specializes other "ge" operations,
bringing it on pair with the current "lt" ops codes.

After applying these changes, invoking the first
check/1 function defined above 30000 times with
different values from 0 to 20 has fallen from 560us
to 370us (measured as average of 3 runs).

### CODE REVIEW

**Comment by josevalim** [comment]
The exact code used for benchmark and an example of the inlined function can be found here: https://gist.github.com/josevalim/6082ce080c7f29312949

**Comment by essen** [comment]
Couple questions:

* Is there an upper limit to the number of clauses created by the inlining?
* Is there still inlining if there are two guard sequences? For example X > 1, X < 5; X > 8, X < 10.
* Wouldn't adding an is_between operation allow for an even greater improvement?

**Comment by josevalim** [comment]
@essen I am not sure your questions apply to this. Can you rephrase them? It is not really generating clauses nor worried about guard sequences. All it does is to move code from one C function into the emulator loop to avoid jumps in the C code. The "inlining" is in moving C code around.

**Comment by josevalim** [comment]
Finally, it is worth pointing out that the select_val approach is still faster, taking around 240us for that test. Therefore it may be worth considering optimizing operations over small intervals into select_val.

For example:

    check(X) when is_integer(X), X >= 0, X <= 20 -> true.

could be transformed into `select_val` and we could do it up to intervals of certain length. The only caveat is that the `is_integer(X)` guard is required.

Another solution for further improvement is adding beam instructions specific for integer checks (if adding more beam instructions is desired). Both suggestions are complementary to this patch.

**Comment by essen** [comment]
OK let me rephrase then.

You optimize:

```
check(X) when X >= 0, X <= 20 -> true.
```

Correct? Does it also optimize:

```
check(X) when X >= 0, X <= 20; X >= 30, X <= 40; X =:= 45 -> true.
```

For example the kind of code you would write to check that a character is [a-z], [A-Z] or $-. (Numbers above inaccurate.)

**Comment by josevalim** [comment]
Yes.

**Comment by essen** [comment]
Fantastic work then! :-)

If you or maintainer need another file to check things, this one should be interesting: https://github.com/ninenines/cowlib/blob/master/src/cow_http_hd.erl

All the macros ?IS_WS, ?IS_ALPHA etc. are similar, though I guess in that case for most of them it's using "or" and checking them one by one. But if *that* isn't optimized the change is easy to make.

"make perfs" can be used to benchmark most of the functions in that file before/after.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by bjorng** [comment]
Well, that comment no longer tells the complete truth. From R12B, floats, lists, and tuples can all be constants.

Perhaps you should just delete that paragraph.

**Comment by proxyles** [comment]
ping  @josevalim

**Comment by bjorng** [comment]
ping @josevalim

**Comment by josevalim** [comment]
I have sent a new pull request for master here: #952. Due to other improvements in master, integer lt/ge comparison is ~60% faster compared to Erlang 18. The new pull request also inlines floats and makes them about ~40% compared to Erlang 18 (about ~25% compared to master itself). :tada:

Feel free to close this one as it is effectively outdated compared to #952.

**Comment by proxyles** [comment]
Closed as per request
